ember-resources/src/util/function.ts

```diff
@@ -92,43 +93,53 @@ export function trackedFunction<Return>(...passedArgs: UseFunctionArgs<Return>)
     fn = passedArgs[2];
   }
 
-  return resource<State<Return>>(context, (hooks) => {
-    let state = new State(initialValue);
+  return resource<TrackedFunctionProperty<Return>>(context, (hooks) => {
+    return new TrackedFunctionProperty(fn, hooks, initialValue);
+  });
+}
 
-    (async () => {
-      try {
-        let notQuiteValue = fn(hooks);
-        let promise = Promise.resolve(notQuiteValue);
+export class TrackedFunctionProperty<Value> {
+  @tracked state: State<Value>;
 
-        waitForPromise(promise);
+  constructor(private fn: ResourceFn<Value>, private hooks: Hooks, initialValue?: Value) {
+    this.state = new State(initialValue);
+    this.getValue();
+  }
 
-        let result = await promise;
+  async getValue() {
+    try {
+      let notQuiteValue = this.fn(this.hooks);
+      let promise = Promise.resolve(notQuiteValue);
 
-        state.error = undefined;
-        state.resolvedValue = result;
-      } catch (e) {
-        state.error = e;
-      } finally {
-        state.isResolved = true;
-      }
-    })();
+      waitForPromise(promise);
 
-    return state;
-  });
-}
+      let result = await promise;
 
-/**
- * State container that represents the asynchrony of a `trackedFunction`
- */
-export class State<Value> {
-  @tracked isResolved = false;
-  @tracked resolvedValue?: Value;
-  @tracked error?: unknown;
+      this.state.error = undefined;
+      this.state.resolvedValue = result;
+    } catch (e) {
+      this.state.error = e;
+    } finally {
+      this.state.isResolved = true;
+    }
+  }
 
-  constructor(public initialValue?: Value) {}
+  @action
+  execute() {
```

Comment:
why "execute" for the name?

Reply:
I'm also not happy with it but `perform` is no kind of reserved for tasks, `run`, `call` etc also seem a bit off and retry is imo trying something that has failed again. Whats your opinion?